### PR TITLE
Fix #568: loading the tabs module failed

### LINKED ISSUE #568: Vimperator don't work in Firefox Beta. Not e10s.
##### Issue type:

- Version compatibility
##### Version:

```
Vimperator 3.15, Nightly 51.0a2, e10s disabled.
```
##### Description:

Mappings and commandline don't work.
In browser console it's showing that `ReferenceError: tabs is not defined`
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Not_defined

### PR DESCRIPTION
**Pull request by ma8ma**
This PR makes to fix #568 loading the tabs module (see also [the candidate](https://github.com/ma8ma/vimperator-labs/pull/1)).

* Replace legacy generator with ES2015 generator for almost places.
* Replace non-standard Array comprehension with Array methods in muttator/content/mail.js.

Several legacy generators still keep because they have been handled by legacy iterators.
* iter() in common/content/base.js
* callbacks of addPageInfoSection() in common/content/buffer.js
* evaluateXPath() in common/content/util.js
* range() in common/content/util.js
* itervalues() in common/content/util.js
* iteritems() in common/content/util.js

### CODE REVIEW

File: muttator/content/mail.js [hunk 1]
```diff
@@ -288,7 +287,9 @@ const Mail = Module("mail", {
      */
     getFoldersWithFlag: function (flag) {
         if (flag) {
-            return [row._folder for ([row] in this.getAllFolderRowMap(true,true)) if (row._folder.flags & flag)]
+            return Array.from(this.getAllFolderRowMap(true, true), ([row]) => row)
+                        .filter(row => row._folder.flags & flag)
+                        .map(row => row._folder);
```
Comment (ma8ma) [inline muttator/content/mail.js]: more short method chains: `Array.from(<generator>, ([row]) => row._folder).filter(folder => folder.flags & flag)`

**Comment by iGEL** [comment]
I'm not qualified to review this, but I want to thank @ma8ma for the work and hope to see a release with this soon :heart: :grin:

File: muttator/content/mail.js [hunk 2]
```diff
@@ -211,20 +211,20 @@ const Mail = Module("mail", {
...
+        function* walkChildren(children, prefix) {
             for (let [, row] in Iterator(children)) {
                 let folder = row._folder;
                 let folderString = prefix + "/" +
                                    (row.useServerNameOnly ? folder.server.prettyName : folder.abbreviatedName);
                 yield [row, folderString];
                 if (row.children.length > 0)
-                    for (let child in walkChildren(row.children, folderString))
+                    for (let child of walkChildren(row.children, folderString))
                         yield child;
```
Comment (ma8ma) [inline muttator/content/mail.js]: unnecessary `for of` loop, can delegate using `yield* walkChildren(...);`

File: muttator/content/mail.js [hunk 3]
```diff
@@ -1179,7 +1179,12 @@ const Mail = Module("mail", {
                     services.get("smtpService").defaultServer = server;
                     return value;
                 },
-                completer: function (context) [[s.key, s.serverURI] for ([, s] in Iterator(mail.smtpServers))],
+                completer: function (context) {
+                    return Array.from((function* () {
+                        for (let [, s] in Iterator(mail.smtpServers))
```
Comment (aureooms) [inline muttator/content/mail.js]: Why keep the old `in Iterator(...)`?

File: common/content/tabs.js [hunk 4]
```diff
@@ -49,10 +49,10 @@ const Tabs = Module("tabs", {
     },
 
     /**
-     * @property {Iterator(Object)} A genenerator that returns all browsers
+     * @return {Generator} A genenerator that returns all browsers
      *     in the current window.
      */
-    get browsers() {
+    browsers: function* () {
```
Comment (aureooms) [inline common/content/tabs.js]: Could have kept the getter by returning the result of an additional generator function. I am not sure which way is best.

File: muttator/content/mail.js (hunk 3 above)
Comment (ma8ma) [inline muttator/content/mail.js]: Sure, `smtpServers` property is Array type, therefore, can simplify to replace with `return mail.smtpServers.map(s => [s.key, s.serverURI]);`

File: common/content/tabs.js (hunk 4 above)
Comment (ma8ma) [inline common/content/tabs.js]: Yeah. `browsers` can create generator within property to keep getter, I'm going to try.

**Comment by timss** [comment]
Thank you so much for this patch.   

Sorry for the delay, but I wasn't able to test this myself until now. I'd also like to reiterate for anyone reading this that Vimperator is (and has been for some time) in need for contributors. Looking at all the (breaking) [changes Mozilla keeps pushing](https://developer.mozilla.org/en-US/docs/Web/JavaScript/New_in_JavaScript/Firefox_JavaScript_changelog#Firefox_51), this certainly will not be the last time. Please see the following links for more information:

- https://github.com/vimperator/vimperator-labs/issues/211#issuecomment-249181301
- https://github.com/vimperator/vimperator-labs/issues/568#issuecomment-274746690

Until a new release has been made, submitted to AMO and finally approved by Mozilla, users will have to build their own versions. Either go for an [unbranded](https://wiki.mozilla.org/Add-ons/Extension_Signing#Unbranded_Builds) release of Firefox, or sign the addon yourself with a [personal AMO API key](https://github.com/vimperator/vimperator-labs/issues/568#issuecomment-275891186).

**Comment by brianpeiris** [comment]
@timss First of all, my thanks to you and @gkatsev for keeping Vimperator alive. Despite the complaints, I'm grateful that the maintainers do respond and update AMO when needed, even if it's not always timely.

> I'd also like to reiterate for anyone reading this that Vimperator is (and has been for some time) in need for contributors.

Wouldn't it help if this was made more obvious with a notice on vimperator.org and in the README here?

**Comment by gkatsev** [comment]
@brianpeiris thanks. I'll make sure to make a note on the AMO page about wanting contributors.

@maxauthority would you be able to update the website?

**Pull request merged**